File: 3_arrays_part3/1_search_in_2d_matrix.py
class Solution:
    def searchMatrix(self, matrix: list[list[int]], target: int) -> bool:
        for row in matrix:
            if row[0] <= target <= row[-1]:
                return self.binary_search(row, target)
        return False

    def binary_search(self, row, target) -> bool:
        n = len(row)
        low, high = 0, n-1
        while low <= high:
            mid = (low + high) // 2
            if row[mid] == target:
                return True
            elif row[mid] < target:
                low = mid+1
            else:
                high = mid-1
        return False


"""
Approach2 ->
Do binary search on entire matrix, Trick Formula -> from mid integer, row and column are mid//m and mid%m.

Time Complexity -> O(logn + logm)
"""


class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        n, m = len(matrix), len(matrix[0])
        low, high = 0, m*n - 1
        while low <= high:
            mid = (low + high) // 2
            r,c = mid//m, mid%m
            if matrix[r][c] == target:
                return True
            elif matrix[r][c] < target:
                low = mid + 1
            else:
                high = mid - 1
        return False


            

File: 03_arrays_part3/1_search_in_2d_matrix.rs
impl Solution {
    pub fn search_matrix(matrix: Vec<Vec<i32>>, target: i32) -> bool {
        let n = matrix.len();
        let m = matrix[0].len();

        let (mut low, mut high) = (0, n * m - 1);

        while low <= high {
            let mid = (low + high) / 2;
            let (r, c) = (mid / m, mid % m);
            if matrix[r][c] == target {
                return true;
            } else if matrix[r][c] < target {
                low = mid + 1;
            } else {
                if mid == 0 {
                    break;
                }
                high = mid - 1;
            }
        }
        false
    }
}
